src/libxively/tls/xi_tls_layer.c

```diff
@@ -333,6 +326,13 @@ static xi_state_t recv_handler( void* context, void* data, xi_state_t in_out_sta
     xi_data_desc_t* ret_buffer = layer_data->decoded_buffer;
     layer_data->decoded_buffer = NULL;
 
+    if ( NULL != layer_data->raw_buffer ||
+         xi_bsp_tls_pending( layer_data->tls_context ) > 0 )
```

Comment:
Yes, you are correct - so it does rise the complexity - but is the only way to guarantee to process whole accessible data to read from both libxively and tls libraries. So there is one case where the memory from the tls read buffer may stay untouched until the next message triggers it.